File: problem/ai-battle-series/09-deadline-miracle/src/main.py
import sys


def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    it = iter(data)
    n = int(next(it))
    nums = [int(next(it)) for _ in range(n)]
    tails = []
    for x in nums:
        lo, hi = 0, len(tails)
        while lo < hi:
            mid = (lo + hi) // 2
            if tails[mid] < x:
                lo = mid + 1
            else:
                hi = mid
        if lo == len(tails):
            tails.append(x)
        else:
            tails[lo] = x
    sys.stdout.write(str(len(tails)) + "\n")


if __name__ == "__main__":
    main()

File: problem/ai-battle-series/09-deadline-miracle/src/main.c
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    int n = 0;
    if (scanf("%d", &n) != 1) {
        return 0;
    }
    if (n <= 0) {
        printf("0\n");
        return 0;
    }

    int *nums = (int *)malloc(sizeof(int) * (size_t)n);
    int *tails = (int *)malloc(sizeof(int) * (size_t)n);
    if (!nums || !tails) {
        free(nums);
        free(tails);
        return 0;
    }

    for (int i = 0; i < n; i++) {
        scanf("%d", &nums[i]);
    }

    int len = 0;
    for (int i = 0; i < n; i++) {
        int x = nums[i];
        int lo = 0;
        int hi = len;
        while (lo < hi) {
            int mid = (lo + hi) / 2;
            if (tails[mid] < x) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        if (lo == len) {
            tails[len++] = x;
        } else {
            tails[lo] = x;
        }
    }

    printf("%d\n", len);
    free(nums);
    free(tails);
    return 0;
}
